### PR TITLE
DRAFT: #33446 with alternate test style

### PR DESCRIPTION
**Pull request by wbclark**
Continuing work from https://github.com/theforeman/puppet-pulpcore/pull/227

### CODE REVIEW

**Comment by wbclark** [comment]
NOTE: I updated the parameter documentation to be clearer for users, and also tried a new style for the tests which should make it easier for developers to interpret test output.

File: spec/acceptance/basic_spec.rb [hunk 1]
```diff
@@ -74,9 +74,20 @@ class { 'pulpcore':
     its(:exit_status) { is_expected.to eq 0 }
   end
 
-  describe command("PULP_SETTINGS=/etc/pulp/settings.py pulpcore-manager diffsettings") do
-    its(:stdout) { is_expected.to match(/^USE_NEW_WORKER_TYPE = True/) }
-    its(:exit_status) { is_expected.to eq 0 }
+  describe "Pulpcore Settings" do
+    let(:settings) { command("PULP_SETTINGS=/etc/pulp/settings.py pulpcore-manager diffsettings") }
```
Comment (wbclark) [inline spec/acceptance/basic_spec.rb:78]: As written, this *works* and should give a more descriptive output in the case of test failures (I will break a test shortly to confirm).

That said, I want to make sure really I understand correctly how this behaves -- would the block argument here get evaluated in the `let` block, or does it later get evaluated three times when calling `settings.exit_status`, `settings.stdout`, etc, causing inefficient code execution?

What I'm trying to do may or may not still be possible with `describe command("PULP_SETTINGS=/etc/pulp/settings.py pulpcore-manager diffsettings") do` instead. I am not sure if `its(:stdout) "will a string work here?" do` would function as I intend it do.

From looking at a lot of rspec examples, `expect(...)` seems to be preferred to `it { is_expected.to ... }` which is why I decided to try this. That said, maybe it is proper rspec style in general but not right for serverspec specifically.

**Review by wbclark — COMMENTED** [review]
Thinking about this some more, what I really want to do is deliver a version of this PR which delivers the feature as needed, adhering to the current test style though it may be flawed.

I thought the tests would work in this style, and indeed they do, but I'm simply not confident enough yet with it to rule out drawbacks.

Therefore I will rename this PR and mark it as a draft, and shortly open another version that delivers the same functionality with the current style tests.

**Comment by wbclark** [comment]
OK, the real version of this is https://github.com/theforeman/puppet-pulpcore/pull/230

File: spec/acceptance/basic_spec.rb (hunk 1 above)
Comment (evgeni) [reply spec/acceptance/basic_spec.rb:78]: I like the new grouping of things, and I think `let` should be evaluated once.

As for `expect` vs `is_expected.to` -- we use the later *everywhere*, so I'd prefer to keep that for consistency (and if change, change globally).

File: manifests/init.pp [hunk 2]
```diff
@@ -132,6 +132,10 @@
 #   available, likely results in performance degradation due to I/O blocking and is not recommended in most cases. Modifying this parameter should
 #   be done incrementally with benchmarking at each step to determine an optimal value for your deployment.
 #
+# @param worker_ttl
+#   The number of seconds before a pulpcore worker should be considered lost. If undefined, pulpcore will use its default setting which is 30 seconds as of
+#   pulpcore version 3.14. You should not need to modify this setting unless the application reports workers timing out while they are busy completing tasks.
```
Comment (evgeni) [inline manifests/init.pp:137]: Reading "You should not need to modify this setting unless the application reports workers timing out while they are busy completing tasks." one could think "Yeah, let's set it to 9000 then!", which is also not what we want the users to do.

For `worker_count` we say " Modifying this parameter should be done incrementally with benchmarking at each step to determine an optimal value for your deployment.", should something similar be added here?

File: spec/acceptance/basic_spec.rb (hunk 1 above)
Comment (ekohl) [reply spec/acceptance/basic_spec.rb:78]: What you can do is:
```ruby
describe 'Pulpcore settings' do
  subject { command("PULP_SETTINGS=/etc/pulp/settings.py pulpcore-manager diffsettings") }

  its(:exit_status) { is_expected.to eq 0 }
end
```

This is effectively the same as what was there before.

Comment (wbclark) [reply spec/acceptance/basic_spec.rb:78]: I think I recall trying something like this in the past which failed:

```ruby
describe "Pulpcore settings" do
  subject { command("PULP_SETTINGS=/etc/pulp/settings.py pulpcore-manager diffsettings") }

  its(:exit_status) "exits without error" { is_expected.to eq 0 }
end
```

I will try again to confirm

Comment (wbclark) [reply spec/acceptance/basic_spec.rb:78]: The issue was

```
Failure/Error: __send__(method, file)

SyntaxError:
  /home/wclark/Projects/puppet-pulpcore/spec/acceptance/basic_spec.rb:80: syntax error, unexpected string literal, expecting `end'
      its(:exit_status) "retrieves pulpcore settings" ...
```

So `it` has an optional string argument before the block argument, but `its(:symbol)` can receive only the block.

**Pull request closed without merging**